server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -303,6 +345,74 @@ static boolean execute(SearchContext searchContext,
+        if (Sort.INDEXORDER.equals(sort)) return null;
+        if (sort.getSort().length > 1) return null; // we need only a single sort
+        SortField sortField = sort.getSort()[0];
+
+        // check if this is a field of type Long, that is indexed and has doc values
+        String fieldName = sortField.getField();
+        final MappedFieldType fieldType = searchContext.mapperService().fullName(fieldName);
+        if (fieldType == null) return  null;
+        if (fieldType.typeName() != "long") return null;
+        if (fieldType.indexOptions() == IndexOptions.NONE) return null;
+        if (fieldType.hasDocValues() == false) return null;
+        byte[] minValueBytes =  PointValues.getMinPackedValue(reader, fieldName);
+        byte[] maxValueBytes =  PointValues.getMaxPackedValue(reader, fieldName);
+        if ((maxValueBytes == null) || (minValueBytes == null)) return null; // no values on this shard
+
+        // finally all checks done, query can be rewritten, rewrite
+        long minValue = LongPoint.decodeDimension(minValueBytes, 0);
+        long maxValue = LongPoint.decodeDimension(maxValueBytes, 0);
+        final long origin = (sortField.getReverse()) ? maxValue : minValue;
+        final long pivotDistance = maxValue == minValue ? maxValue/2 : (maxValue - minValue)/2;
+        return LongPoint.newDistanceFeatureQuery(sortField.getField(), 1, origin, pivotDistance);
+    }
+
+    static TopDocs convertToTopFieldDocs(SortField[] sortFields, IndexReader reader, QuerySearchResult result) throws IOException {
+        // 1. sort docIds, so DocValues iterator can iterate over docIds in ascending order
+        // keep their original positions in positions
+        TopDocsAndMaxScore topDocsAndMaxScore = result.topDocs();
+        ScoreDoc[] oldScoreDocs = topDocsAndMaxScore.topDocs.scoreDocs;
+        int[] docIds = new int[oldScoreDocs.length];
+        int[] positions = new int[oldScoreDocs.length];
+        for (int i = 0; i < oldScoreDocs.length; i++) {
+            docIds[i] = oldScoreDocs[i].doc;
+            positions[i] = i;
+        }
+        new InPlaceMergeSorter() {
+            @Override
+            public int compare(int i, int j) {
+                return Integer.compare(docIds[i], docIds[j]);
+            }
+            @Override
+            public void swap(int i, int j) {
+                int tempDocId = docIds[i];
+                docIds[i] = docIds[j];
+                docIds[j] = tempDocId;
+
+                int pos = positions[i];
+                positions[i] = positions[j];
+                positions[j] = pos;
+            }
+        }.sort(0, docIds.length);
+
+        // 2. Iterate over DocValues and fill fieldDocs
+        Long missingValue = (Long) sortFields[0].getMissingValue(); // for rewritten numeric sort, when we had a single sort on Long field
+        SortedNumericDocValues sdvs = MultiDocValues.getSortedNumericValues(reader, sortFields[0].getField());
```

Comment:
@jpountz Is it fine to use `MultiDocValues.getSortedNumericValues` or is there is some other way to get values for docs?

Reply:
I wonder if this could be simplified if we add the numeric field as the secondary sort. This way you can access the values directly in the returned field docs ?